### DISCUSSION TITLE
Any way to modify the root HTML tag in Storybook previews?

### DISCUSSION THREAD

**Discussion by ghost** [storybookjs/storybook#26959] [Help]
### Summary

Storybook provides the `preview-head.html` and `preview-body.html` templates to customise those tags in the rendered previews, but I would like to be able to override the `<html>` tag with data attributes for a React application. Almost like if there was a `preview-root.html` template.

I haven't been able to find a way to configure any attributes on root html tag that happen consistently before the preview is rendered. I've noticed that `lang` attribute is set on it, maybe inherited from the Storybook manager. There doesn't seem to be any mention in the docs of how to work with the document root for setting lang or other attributes.

There are plugins like `storybook-addon-root-attribute` but it seems to add the attributes after the preview has rendered, and then queries the element and sets attrs.

<img width="460" alt="image" src="https://github.com/storybookjs/storybook/assets/142195963/05c35176-275d-4298-9b29-e4d8d8c73b46">

(my question only relates to the `<html>` tag in the preview iframe, not the html root of Storybook manager, the wrapper application)

### Additional information

Storybook v8

**Comment by firxworx** [comment]
I hit the same problem... and the addons that are out there that I tried are broken or require much older versions of storybook.

I tried to add a decorator in my `.storybook/preview.tsx` and I found that in a react `useEffect` (and then later trying outside of one) I simply couldn't read or write attributes of the document element.

## Solution

Thanks to finding your question above I learned you can add `.storybook/preview-head.html` to customize the contents of `head` for the document generated for the preview iframe.

If you add a `<script>...</script>` in there to manipulate the `<html>` tag it will work. Example:

```js
<script>
  (function() {
    document.documentElement.setAttribute('dir', 'ltr');
  })();
</script>
```

Based on your question, that might be all the information you need right there!

### Dynamic Values

I wanted to be able to change locales `lang` and `dir` so I added a toolbar entry and a decorator to `.storybook/preview.tsx`.

To the default preview export / config obect I defined `globalTypes.locale: {...}` to add a toolbar entry for myself that would expose the locale to `context.globals.locale`. This is accessible within decorators because they are passed the `context` object.

I wired it up to a `Decorator` component that has a `useEffect()` that fires custom events whenever the `locale` or derived `direction` changes. 

In the `preview-head.html` script I added listeners for those events and update the document element attributes accordingly.

Storybook itself doesn't seem to be very RTL friendly so the preview pane looks funny in RTL but the important part for me was developing and testing my project's components and this does the trick.

### FYI

When reading further in the docs there does seem to be a way to edit the document itself but it is extremely overkill for this use-case so I went with the Solution explained above. Its what builders have to do:

- https://storybook.js.org/docs/configure/story-rendering
- https://github.com/storybookjs/storybook/blob/next/code/builders/builder-webpack5/templates/preview.ejs

**Reply by andrewpeters9** [reply]
@firxworx Thanks for the write up! Any chance you could share your code for dynamically changing the values?

**Reply by onx2** [reply]
FWIW I'm using a mutation observer to fix my darkmode in deployed versions of storybook. You could use a similar approach to watch for changes to other elements or attributes.

This watches the body element's class list for changes to "dark" class because that is what is used by `storybook-dark-mode` and then updates the iframe's HTML element with the data attribute I need.

```html
<!-- preview-head.html -->
<script src="./.storybook/watchDarkModeInPreview.js"></script>
```

```js
// watchDarkModeInPreview.js
function watchDarkModeInPreview() {
  const init = () => {
    const body = document.body;
    if (!body) return;

    const updateDarkMode = () => {
      if (body.classList.contains("dark")) {
        document.documentElement.setAttribute("data-dark-mode", "true");
      } else {
        document.documentElement.removeAttribute("data-dark-mode");
      }
    };

    const observer = new MutationObserver(updateDarkMode);
    observer.observe(body, { attributeFilter: ["class"] });

    updateDarkMode();
  };

  if (document.readyState !== "loading") {
    init();
  } else {
    document.addEventListener("DOMContentLoaded", init);
  }
}

watchDarkModeInPreview();
```
